drv/transceivers-server/src/net_api.rs

```diff
@@ -0,0 +1,414 @@
+                    out[count] = status.bits();
+                    count += 1;
+                }
+                Ok((HostResponse::Status, count))
+            }
+            HostRequest::Read(mem) => {
+                let out_size = mem.len() as u32 * modules.ports.0.count_ones();
+                if out_size as usize > transceiver_messages::MAX_MESSAGE_SIZE {
+                    return Err(Error::RequestTooLarge);
+                }
+                self.read(mem, modules, out)?;
+                Ok((HostResponse::Read(mem), out_size as usize))
+            }
+            HostRequest::Write(mem) => {
+                let data_size = mem.len() as u32 * modules.ports.0.count_ones();
+                // TODO: check equality here and return a different error?
+                if data_size as usize > data.len() {
+                    return Err(Error::RequestTooLarge);
+                }
+                self.write(mem, modules, data)?;
+                Ok((HostResponse::Write(mem), 0))
+            }
+            HostRequest::SetPowerMode(mode) => {
+                let mask = get_mask(modules)?;
+                // TODO: do we need delays in between any of these operations?
+                match mode {
+                    PowerMode::Off => {
+                        // Power disabled, LpMode enabled (the latter doesn't
+                        // make a difference, but keeps us in a known state)
+                        self.transceivers
+                            .clear_power_enable(mask)
+                            .map_err(|_e| Error::ReadFailed)?;
+                        self.transceivers
+                            .set_reset(mask)
+                            .map_err(|_e| Error::ReadFailed)?;
+                        self.transceivers
+                            .set_lpmode(mask)
+                            .map_err(|_e| Error::ReadFailed)?;
+                    }
+                    PowerMode::Low => {
+                        // Power enabled, LpMode enabled
+                        self.transceivers
+                            .set_lpmode(mask)
+                            .map_err(|_e| Error::ReadFailed)?;
+                        self.transceivers
+                            .set_power_enable(mask)
+                            .map_err(|_e| Error::ReadFailed)?;
+                        self.transceivers
+                            .clear_reset(mask)
+                            .map_err(|_e| Error::ReadFailed)?;
+                    }
+                    PowerMode::High => {
+                        // Power enabled, LpMode disabled
+                        self.transceivers
+                            .clear_lpmode(mask)
+                            .map_err(|_e| Error::ReadFailed)?;
+                        self.transceivers
+                            .set_power_enable(mask)
+                            .map_err(|_e| Error::ReadFailed)?;
```

Comment:
LpMode really only seems to matter when the module is both powered on and out of reset. Given that, changing it prior to those conditions seems fine.

Also see my comment above about the FPGA handling these details in the near future, not the SP. So while in the mode we are doing `set_power_enable` after `clear_lpmode`, the power should already be enabled as part of `PowerMode::Low` (aka A2).

Reply:
Thanks Aaron, that's helpful. Does that mean that the SP will not even issue multiple commands like this? That instead, there'd be a requested transition from say A2 -> A0, and the FPGA would set these pins?

Reply:
Correct, it would issue a single "go to power state `X`" command. The intention here will be for the SP to request a module to get to a certain state and then the FPGA would handle all the signal-level details of what that means. If the FPGA were to fail to comply with the request (due to a power fault for example), it will report as such so when the SP realizes that the port is not in the state it requested it could go get details as to why that is the case.